From 0400f1d2e9b42b7e5eec011575700b1d45f7bea6 Mon Sep 17 00:00:00 2001
From: Sarika Sharma <quic_sarishar@quicinc.com>
Date: Mon, 10 Jun 2024 10:14:58 +0530
Subject: [PATCH] wifi: ath12k: add WMI service bit for WDS null frame support

Changes are done to check WMI_SERVICE_WDS_NULL_FRAME_SUPPORT service bit
set to know that WDS null packet feature is supported by firmware.
If WDS null frame is supported by firmware, send resource config flag2
WMI_RSRC_CFG_FLAGS2_WDS_NULL_FRAME_SUPPORT(bit 22) as set to acknowledge
firmware that host is using WDS null frame feature.

Currently, the 4addr NULL frame packets are consumed by firmware and
not handed over to host for any processing. In WDS mode
station would be sending 4address NULL frame during assoc
succes case and AP uses that frame to create AP_VLAN interface.

This feature is required by host to allow AP to create AP_VLAN interface
in WDS mode.

Patch-dependency: 899-ath12k-Add-support-for-WDS-offload-enable.patch

Signed-off-by: Sarika Sharma <quic_sarishar@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/wmi.c | 7 +++++++
 drivers/net/wireless/ath/ath12k/wmi.h | 3 +++
 2 files changed, 10 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -5969,6 +5969,9 @@ ath12k_wmi_copy_resource_config(struct a
 	else
 		wmi_cfg->flags2 |= WMI_RSRC_CFG_FLAGS2_FW_AST_INDICATION_DISABLE;
 
+	if (tg_cfg->is_wds_null_frame_supported)
+		wmi_cfg->flags2 |= WMI_RSRC_CFG_FLAGS2_WDS_NULL_FRAME_SUPPORT;
+
 #ifdef CPTCFG_ATH12K_SAWF
 	wmi_cfg->flags2 |= (tg_cfg->sawf & ath12k_sawf_enable) ?
 			   (WMI_RSRC_CFG_FLAGS2_SAWF_CONFIG_ENABLE_SET) : (0);
@@ -6197,6 +6200,8 @@ int ath12k_wmi_cmd_init(struct ath12k_ba
 		arg.res_cfg.is_reg_cc_ext_event_supported = true;
 	if (test_bit(WMI_TLV_SERVICE_RADAR_FLAGS_SUPPORT, ab->wmi_ab.svc_map))
 		arg.res_cfg.is_full_bw_nol_feature_supported = true;
+	if (test_bit(WMI_SERVICE_WDS_NULL_FRAME_SUPPORT, ab->wmi_ab.svc_map))
+		arg.res_cfg.is_wds_null_frame_supported = true;
 
 	ab->hw_params->wmi_init(ab, &arg.res_cfg);
 
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -2453,6 +2453,7 @@ enum wmi_tlv_service {
 	WMI_TLV_SERVICE_RADAR_FLAGS_SUPPORT = 390,
 	WMI_TLV_SERVICE_SW_PROG_DFS_SUPPORT = 395,
 	WMI_TLV_SERVICE_THERM_THROT_POUT_REDUCTION = 410,
+	WMI_SERVICE_WDS_NULL_FRAME_SUPPORT = 421,
 	WMI_MAX_EXT2_SERVICE,
 };
 
@@ -2763,6 +2764,7 @@ struct ath12k_wmi_resource_config_arg {
 	u8  dp_peer_meta_data_ver;
 	bool sawf;
 	bool is_full_bw_nol_feature_supported;
+	bool is_wds_null_frame_supported;
 };
 
 struct ath12k_wmi_init_cmd_arg {
@@ -2856,6 +2858,7 @@ struct wmi_init_cmd {
 #define WMI_RSRC_CFG_FLAGS2_CALC_NEXT_DTIM_COUNT_SET      BIT(9)
 #define WMI_RSRC_CFG_FLAGS2_INTRABSS_MEC_WDS_LEARNING_DISABLE	BIT(15)
 #define WMI_RSRC_CFG_FLAGS2_FW_AST_INDICATION_DISABLE		BIT(18)
+#define WMI_RSRC_CFG_FLAGS2_WDS_NULL_FRAME_SUPPORT		BIT(22)
 #define WMI_RSRC_CFG_FLAG1_THREE_WAY_COEX_CONFIG_OVERRIDE_SUPPORT BIT(25)
 #define WMI_RSRC_CFG_FLAG1_BSS_CHANNEL_INFO_64	BIT(5)
 #define WMI_RSRC_CFG_FLAGS2_SAWF_CONFIG_ENABLE_SET	  BIT(13)
